File: src/arguments.cpp
#include <map>
#include <string>
#include <iostream>
#include <cstdlib>

/*
 * Accept arguments with pattern like "cmd --param0 value0 --param1 value1".
 */

int main(int argc, char* argv[]) {
    
    std::map<std::string, std::string> arguments;
    
    if (argc % 2 == 0) {
        std::cerr << "improper arguments";
        return EXIT_FAILURE;
    }
    
    if (argc > 1) {
        
        for (int i = 1; i < argc - 1; i += 2) {
            std::string paramter_t(argv[i]);
            if (paramter_t.substr(0, 2) != "--"){
                std::cerr << "improper arguments";
                return EXIT_FAILURE;
            }
            
            std::string parameter = paramter_t.substr(2, paramter_t.length() - 2);
            std::string value(argv[i + 1]);
            
            arguments.insert(std::pair<std::string, std::string>(parameter, value)); //Alternatively: arguments[parameter] = value;
        }
    }
    
    std::cout << "{";
    std::size_t lastCheck = 0;
    for (auto iter = arguments.begin(); iter != arguments.end(); iter++) {
        lastCheck ++;
        std::cout << "\"" + iter->first + "\": \"" + iter->second + "\"";
        if (lastCheck != arguments.size()) {
            std::cout << ", ";
        }
    }
    std::cout << "}";
    
    return EXIT_SUCCESS;
}

File: src/arguments.rs
//! Read arguments from shell

// declare modules

// imports
use std::env;
use std::collections::HashMap;


/// Main method: use standard lib to parse arguments.
/// Accept arguments with pattern like "cmd --param0 value0 --param1 value1".
pub fn main() {

    // result
    let mut arguments: HashMap<&str, &String> = HashMap::new();

    // parse arguments
    let argv: Vec<String> = env::args().collect();
    let argc: usize = argv.len();


    if argc % 2 == 0 {
        eprintln!("improper arguments");
        panic!();
    }

    if argc > 1 {
        for _i in (1 .. argc).step_by(2) {
            if let Some(paramter) = argv[_i].strip_prefix("--") {
                arguments.insert(paramter, &argv[_i + 1]);


            } else {
                eprintln!("improper arguments");
                panic!();
            }
        }
    }

    // print the entire HashMap using debug formatting
    println!("{:?}", arguments);
}
